### PR TITLE
[BEAM-14486] Document pubsubio & fix its behavior.

### PR DESCRIPTION
**Pull request by lostluck**
Removes experimental note from the Go SDK's PubSubio and make it clear that it only works with Dataflow as currently implemented. If that changes, so too can the documentation.

Fix a behavioral difference with the python implementation, where raw bytes messages weren't wrapped in messages before writing. They are now wrapped. Explicitly panic for other PCollection inputs.

Since this only works on dataflow as implemented, these changes have been manually tested with a modified wordcap that sends data through a PubSub loop (publishing to the same topic being read from). Each pipeline was then successfully drained.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @jrmccluskey

File: sdks/go/pkg/beam/io/pubsubio/pubsubio.go [hunk 1]
```diff
@@ -73,15 +80,38 @@ func Read(s beam.Scope, project, topic string, opts *ReadOptions) beam.PCollecti
 	return out[0]
 }
 
-func unmarshalMessageFn(raw []byte) (*pb.PubsubMessage, error) {
+func unmarshalMessageFn(raw []byte, emit func(*pb.PubsubMessage)) error {
```
Comment (damccorm) [inline sdks/go/pkg/beam/io/pubsubio/pubsubio.go:83]: This emitter should be registered

File: sdks/go/pkg/beam/io/pubsubio/pubsubio.go [hunk 2]
```diff
@@ -73,15 +80,38 @@ func Read(s beam.Scope, project, topic string, opts *ReadOptions) beam.PCollecti
...
+}
+
+func wrapInMessage(raw []byte, emit func(*pb.PubsubMessage)) {
+	emit(&pb.PubsubMessage{
+		Data: raw,
+	})
 }
 
-// Write writes PubSubMessages or bytes to the given pubsub topic.
+func marshalMessageFn(in *pb.PubsubMessage, emit func([]byte)) error {
```
Comment (damccorm) [inline sdks/go/pkg/beam/io/pubsubio/pubsubio.go:98]: Same comment, this emitter should be registered

**Review by jrmccluskey — APPROVED** [review]
LGTM mod Danny's comments around generics registration. An integration test would be nice but  I understand the limitations around it.

**Pull request merged**